File: api/src/services/omdb.rs
use std::sync::Arc;

use crate::error::AppError;
use crate::services::api_key_pool::ApiKeyPool;
use crate::services::retry::{self, OMDB_RETRY};
use serde::Deserialize;

#[derive(Clone)]
pub struct OmdbClient {
    key_pool: Arc<ApiKeyPool>,
    http: reqwest::Client,
}

#[derive(Debug, Deserialize)]
pub struct OmdbResponse {
    #[serde(rename = "Ratings", default)]
    pub ratings: Vec<OmdbRating>,
    #[serde(rename = "imdbRating")]
    pub imdb_rating: Option<String>,
    #[serde(rename = "Metascore")]
    pub metascore: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct OmdbRating {
    #[serde(rename = "Source")]
    pub source: String,
    #[serde(rename = "Value")]
    pub value: String,
}

impl OmdbClient {
    pub fn new(key_pool: ApiKeyPool, http: reqwest::Client) -> Self {
        Self { key_pool: Arc::new(key_pool), http }
    }

    pub async fn get_ratings(&self, imdb_id: &str) -> Result<OmdbResponse, AppError> {
        let api_key = self.key_pool.active_key_raw();
        let imdb_id = imdb_id.to_owned();
        let resp = retry::send_with_retry(&OMDB_RETRY, || {
            self.http
                .get("https://www.omdbapi.com/")
                .query(&[("apikey", api_key.as_str()), ("i", imdb_id.as_str())])
                .send()
        })
        .await?
        .error_for_status()?;
        Ok(resp.json().await?)
    }
}

File: api-go/internal/services/omdb.go
package services

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"

	"openposterdb/internal/errors"
)

type OmdbClient struct {
	KeyPool *APIKeyPool
	HTTP    *http.Client
}

func NewOmdbClient(keys []string, httpClient *http.Client) *OmdbClient {
	return &OmdbClient{
		KeyPool: NewAPIKeyPool(keys),
		HTTP:    httpClient,
	}
}

type OmdbResponse struct {
	Ratings    []OmdbRating `json:"Ratings"`
	IMDBRating *string      `json:"imdbRating"`
	Metascore  *string      `json:"Metascore"`
}

type OmdbRating struct {
	Source string `json:"Source"`
	Value  string `json:"Value"`
}

func (c *OmdbClient) GetRatings(imdbID string) (*OmdbResponse, error) {
	apiKey := c.KeyPool.ActiveKeyRaw()
	url := fmt.Sprintf("https://www.omdbapi.com/?apikey=%s&i=%s", apiKey, imdbID)

	resp, err := SendWithRetry(&OMDBRetry, func() (*http.Response, error) {
		return c.HTTP.Get(url)
	})
	if err != nil {
		return nil, errors.NewAPIError(err)
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	if resp.StatusCode != http.StatusOK {
		return nil, errors.NewOther(fmt.Sprintf("OMDb returned %d: %s", resp.StatusCode, string(body)))
	}

	var result OmdbResponse
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, errors.NewAPIError(err)
	}
	return &result, nil
}
